### PR TITLE
Message Forwarding

### PR DESCRIPTION
**Pull request by DA-344**
## Summary

[redmvc said in #9892](https://github.com/Rapptz/discord.py/pull/9892#issuecomment-2380867248) they will not be able to continue with the PR anytime soon so I've continued their work in this PR.

This has all the previous changes in addition of:
- `PartialMessage.forward(...)` method for message forwarding
- `MessageReference` constructor not a breaking change
- Moved `MessageSnapshot` from "Data Classes" section to "Discord Models"
- Added `MessageReferenceType` enum to docs
- `.. versionadded::`s where it was missing.

**Related Discord API Docs PRs**:
- https://github.com/discord/discord-api-docs/pull/6833
- https://github.com/discord/discord-api-docs/pull/7116

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/enums.py [hunk 1]
```diff
@@ -218,6 +218,13 @@ def __str__(self) -> str:
         return self.name
 
 
+class MessageReferenceType(Enum):
+    reply = 0
+    forward = 1
+
+    default = 0
```
Comment (Rapptz) [inline discord/enums.py:225]:
```suggestion
class MessageReferenceType(Enum):
    default = 0
    reply = 0
    forward = 1
```

**Review by Rapptz — COMMENTED** [review]
This looks good but there are unfortunately merge conflicts.

File: discord/message.py [hunk 2]
```diff
@@ -1768,6 +1971,10 @@ class Message(PartialMessage, Hashable):
         The poll attached to this message.
 
         .. versionadded:: 2.4
+    message_snapshots: Optional[List[:class:`MessageSnapshot`]]
```
Comment (Rapptz) [inline discord/message.py:1974]:
```suggestion
    message_snapshots: List[:class:`MessageSnapshot`]
```

`Optional[List[X]]` is silly.

**Review by Rapptz — CHANGES_REQUESTED** [review]
Sorry, I forgot this PR also introduces instances of `Optional[List[X]]`.

File: discord/message.py (hunk 2 above)
Comment (DA-344) [reply discord/message.py]: Updated on the latest commit.

**Comment by DA-344** [comment]
Merge conflicts have been resolved and suggestions have been commited, PR is now ready to be merged.

**Comment by Rapptz** [comment]
This has conflicts again.

**Comment by DA-344** [comment]
Fixed merge conflicts again.

**Pull request merged**